### PR TITLE
[training] freeze decoder

### PR DESCRIPTION
**Pull request by eustlb**
Add the possibility to freeze the decoder. Note that freezing the decoder will freeze the decoder `embed_tokens` layer that is by default tied to `proj_out` layers (not a layer of the decoder). This way, `proj_out` also gets frozen, it is thus necessary to unfreeze it. 

other minor changes:
- use `set_seed` [from Accelerate](https://huggingface.co/docs/accelerate/v0.1.0/internal.html#accelerate.utils.set_seed) that will also set the seed for numpy
- add naming of wandb runs

### CODE REVIEW

File: training/run_distillation.py [hunk 1]
```diff
@@ -976,6 +999,13 @@ def set_trainable_parameters(module, requires_grad=False):
     if training_args.freeze_encoder:
         set_trainable_parameters(student_model.model.encoder, requires_grad=False)
         student_model.model.encoder.gradient_checkpointing = False
+    
+    if training_args.freeze_decoder:
+        set_trainable_parameters(student_model.model.decoder, requires_grad=False)
+        student_model.model.decoder.gradient_checkpointing = False
+        # un-freeze LM head parameters (and consequently word embeddings), frozen when frozing decoder since tied word embedding and LM head
+        for p in student_model.proj_out.parameters():
+            p.requires_grad = True
```
Comment (sanchit-gandhi) [inline training/run_distillation.py:1008]:
```suggestion
        # un-freeze LM head parameters (and consequently word embeddings), frozen when freezing decoder since tied word embedding and LM head
        set_trainable_parameters(student_model.model.encoder, requires_grad=True)
```

File: training/run_distillation.py [hunk 2]
```diff
@@ -350,6 +358,14 @@ class DistillationTrainingArguments(Seq2SeqTrainingArguments):
             )
         },
     )
+    freeze_decoder: Optional[bool] = field(
+        default=False,
+        metadata={
+            "help": (
+                "Whether to freeze the entire decoder model."
```
Comment (sanchit-gandhi) [inline training/run_distillation.py:365]:
```suggestion
                "Whether to freeze the entire decoder model. Note that the decoder input embeddings are **not** frozen, since they are tied to the LM head."
```

File: training/run_distillation.py [hunk 3]
```diff
@@ -37,6 +37,7 @@
 import transformers
 from accelerate import Accelerator
 from accelerate.logging import get_logger
+from accelerate.utils import set_seed
```
Comment (sanchit-gandhi) [inline training/run_distillation.py:40]: Nice catch!

File: training/run_distillation.py [hunk 4]
```diff
@@ -1315,10 +1349,18 @@ def compute_metrics(preds, labels):
         eval_steps = training_args.eval_steps
 
     # 13. Define optimizer, LR scheduler, collator
+    
+    if training_args.freeze_encoder:
+        forbidden_module = [student_model.model.encoder]
+        if training_args.freeze_decoder:
```
Comment (sanchit-gandhi) [inline training/run_distillation.py:1355]: Proposed code would work for most cases, but would miss the edge case when `--freeze_encoder=False` and `--freeze_decoder=True`

**Review by sanchit-gandhi — APPROVED** [review]
Looks good - just some minor comments

File: training/run_distillation.py (hunk 4 above)
Comment (eustlb) [reply training/run_distillation.py]: Spot on! fixed in [b33e7f8](https://github.com/huggingface/distil-whisper/pull/126/commits/b33e7f8c0b8525dd9dafdb5c13b8adb129d17b84)kljl

File: training/run_distillation.py (hunk 1 above)
Comment (eustlb) [reply training/run_distillation.py]: If think you meant  `set_trainable_parameters(student_model.proj_out, requires_grad=True)`. Implemented in [7280623](https://github.com/huggingface/distil-whisper/pull/126/commits/7280623e6bfd8343e455fc465046c3fd64b4f8fd)
